docs/reference/pip_install.rst

```diff
@@ -21,9 +21,52 @@ Overview
 
 Pip install has several stages:
 
-1. Resolve dependencies. What will be installed is determined here.
-2. Build wheels. All the dependencies that can be are built into wheels.
-3. Install the packages (and uninstall anything being upgraded/replaced).
+1. Identify the base requirements. The user supplied arguments are processed
+   here.
+2. Resolve dependencies. What will be installed is determined here.
+3. Build wheels. All the dependencies that can be are built into wheels.
+4. Install the packages (and uninstall anything being upgraded/replaced).
+
+Argument Handling
++++++++++++++++++
+
+When looking at the items to be installed, pip checks what type of item
+each is, in the following order:
+
+1. Project or archive URL.
+2. Local directory (which must contain a ``setup.py``, or pip will report
+   an error).
+3. Local file (a sdist or wheel format archive, following the naming
+   conventions for those formats).
+4. A requirement, as specified in PEP 440.
+
+Each item identified is added to the set of requirements to be satisfied by
+the install.
+
+Working Out the Name and Version
+++++++++++++++++++++++++++++++++
+
+For each candidate item, pip needs to know the project name and version. For
+wheels (identified by the ``.whl`` file extension) this can be obtained from
+the filename, as per the Wheel spec. For local directories, or explicitly
+specified sdist files, the ``setup.py egg_info`` command is used to determine
+the project metadata. For sdists located via an index, the filename is parsed
+for the name and project version (this is in theory slightly less reliable
+than using the ``egg_info`` command, but avoids downloading and processing
+unnecessary numbers of files).
+
+Any URL may use the ``#egg`` syntax (see :ref:`VCS Support`) to explicitly
+state the project name and version.
```

Comment:
OK, that's even messier than I'd spotted (I'd seen the check from your second bullet but hadn't noticed that it was inconsistent over the format to expect). But I think that ```#egg=ProjectName``` should work everywhere, so maybe I'll reword this paragraph to say

> Any URL may use the ``#egg=name`` syntax (see :ref:`VCS Support`) to explicitly state the project name.